// @glimmer/runtime/dist/commonjs/lib/compiled/expressions/args.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const positional_args_1 = require("./positional-args");
exports.COMPILED_EMPTY_POSITIONAL_ARGS = positional_args_1.COMPILED_EMPTY_POSITIONAL_ARGS;
exports.CompiledPositionalArgs = positional_args_1.CompiledPositionalArgs;
exports.EvaluatedPositionalArgs = positional_args_1.EvaluatedPositionalArgs;
const named_args_1 = require("./named-args");
exports.COMPILED_EMPTY_NAMED_ARGS = named_args_1.COMPILED_EMPTY_NAMED_ARGS;
exports.CompiledNamedArgs = named_args_1.CompiledNamedArgs;
exports.EvaluatedNamedArgs = named_args_1.EvaluatedNamedArgs;
const reference_1 = require("@glimmer/reference");
exports.EMPTY_BLOCKS = {
    default: null,
    inverse: null
};
class CompiledArgs {
    constructor(positional, named, blocks) {
        this.positional = positional;
        this.named = named;
        this.blocks = blocks;
        this.type = "compiled-args";
    }
    static create(positional, named, blocks) {
        if (positional === positional_args_1.COMPILED_EMPTY_POSITIONAL_ARGS && named === named_args_1.COMPILED_EMPTY_NAMED_ARGS && blocks === exports.EMPTY_BLOCKS) {
            return this.empty();
        }
        else {
            return new this(positional, named, blocks);
        }
    }
    static empty() {
        return exports.COMPILED_EMPTY_ARGS;
    }
    evaluate(vm) {
        let { positional, named, blocks } = this;
        return EvaluatedArgs.create(positional.evaluate(vm), named.evaluate(vm), blocks);
    }
}
exports.CompiledArgs = CompiledArgs;
exports.COMPILED_EMPTY_ARGS = new (class extends CompiledArgs {
    constructor() {
        super(positional_args_1.COMPILED_EMPTY_POSITIONAL_ARGS, named_args_1.COMPILED_EMPTY_NAMED_ARGS, exports.EMPTY_BLOCKS);
    }
    evaluate(_vm) {
        return EMPTY_EVALUATED_ARGS;
    }
});
class EvaluatedArgs {
    constructor(positional, named, blocks) {
        this.positional = positional;
        this.named = named;
        this.blocks = blocks;
        this.tag = reference_1.combineTagged([positional, named]);
    }
    static empty() {
        return EMPTY_EVALUATED_ARGS;
    }
    static create(positional, named, blocks) {
        return new this(positional, named, blocks);
    }
    static positional(values, blocks = exports.EMPTY_BLOCKS) {
        return new this(positional_args_1.EvaluatedPositionalArgs.create(values), named_args_1.EVALUATED_EMPTY_NAMED_ARGS, blocks);
    }
    static named(map, blocks = exports.EMPTY_BLOCKS) {
        return new this(positional_args_1.EVALUATED_EMPTY_POSITIONAL_ARGS, named_args_1.EvaluatedNamedArgs.create(map), blocks);
    }
}
exports.EvaluatedArgs = EvaluatedArgs;
const EMPTY_EVALUATED_ARGS = new EvaluatedArgs(positional_args_1.EVALUATED_EMPTY_POSITIONAL_ARGS, named_args_1.EVALUATED_EMPTY_NAMED_ARGS, exports.EMPTY_BLOCKS);
